### PR TITLE
Bug 1647458 - Bulk action handling / Todo list

### PR DESCRIPTION
**Pull request by SuyashSalampuria**

### CODE REVIEW

**Comment by SuyashSalampuria** [comment]
@camd I need some help in here
I see that get request are working fine but the post and and delete say that its not authorized.
Do I need to pass some headers or something such here
I tried even using postman but got same errors on post request in the API
Work is still in progress here although

**Comment by camd** [comment]
Were you logged in when you tried?  This will require a log in.  

When I try this and am logged in, I get this error:

     Traceback (most recent call last):
            |   File "/usr/local/lib/python3.7/site-packages/django/core/handlers/exception.py", line 34, in inner
            |     response = get_response(request)
            |   File "/usr/local/lib/python3.7/site-packages/django/core/handlers/base.py", line 115, in _get_response
            |     response = self.process_exception_by_middleware(e, request)
            |   File "/usr/local/lib/python3.7/site-packages/django/core/handlers/base.py", line 113, in _get_response
            |     response = wrapped_callback(request, *callback_args, **callback_kwargs)
            |   File "/usr/local/lib/python3.7/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
            |     return view_func(*args, **kwargs)
            |   File "/usr/local/lib/python3.7/site-packages/rest_framework/viewsets.py", line 114, in view
            |     return self.dispatch(request, *args, **kwargs)
            |   File "/usr/local/lib/python3.7/site-packages/rest_framework/views.py", line 505, in dispatch
            |     response = self.handle_exception(exc)
            |   File "/usr/local/lib/python3.7/site-packages/rest_framework/views.py", line 465, in handle_exception
            |     self.raise_uncaught_exception(exc)
            |   File "/usr/local/lib/python3.7/site-packages/rest_framework/views.py", line 476, in raise_uncaught_exception
            |     raise exc
            |   File "/usr/local/lib/python3.7/site-packages/rest_framework/views.py", line 502, in dispatch
            |     response = handler(request, *args, **kwargs)
            |   File "/usr/local/lib/python3.7/site-packages/rest_framework/mixins.py", line 19, in create
            |     self.perform_create(serializer)
            | TypeError: perform_create() missing 1 required positional argument: 'project'

**Comment by SuyashSalampuria** [comment]
@camd I have switched to using set for storing the tests which have been selected because it can delete tests by their value and hence prevent searching to delete inside a normal array
But, I see that storing and deleting the tests using the method I have used is too slow
It would be great if you could review it once
This feature is almost ready except for some finishing chnages

**Comment by camd** [comment]
@SuyashSalampuria : I will review this today.  Thanks for your hard work.  I'm excited to see this feature land!

File: treeherder/push_health/tests.py [hunk 1]
```diff
@@ -123,6 +123,16 @@ def get_current_test_failures(push, option_map, jobs):
         isClassifiedIntermittent = any(
             job['failure_classification_id'] == 4 for job in jobs[job_name]
         )
+        is_investigated = any(
```
Comment (camd) [inline treeherder/push_health/tests.py:126]: I would think this would work and be quick.  Is this what is too slow?  I wonder if you first group all `investigated_tests` by the `job_type`?  Then you could check any test to see if it's in that `dict` and check inside it if it has a matching `test_name`?

Comment (SuyashSalampuria) [reply treeherder/push_health/tests.py:126]: No, this is not at all slow. 
I was speaking of things in frontend
When we check against any test, it goes on to add that test in the state variable in selectedTests (in the Test.jsx file)
I see that the check is not done instantly but took a bit of time on my PC

> Another approach would be do download the investigated tests for the push and match them up in the UI/JavaScript

Should I try doing it like this. This would need another API call and I see no big advantage of this too

ALso, for making the things real time ( disabling the tests as soon as it marked investigated) , should I make anothet get request to the push health page or manually add things to the state.

**Comment by SuyashSalampuria** [comment]
ALso, I don't know why the content of text_extract_job.sql have changed so much
Is it whenever a new migration is applied, it changes.I see huge diff in it now
Should I commit the new test_extract_job.sql file too

File: treeherder/push_health/tests.py (hunk 1 above)
Comment (camd) [reply treeherder/push_health/tests.py:126]: @SuyashSalampuria : I would not re-fetch the data.  I would update the test on the front-end when you get a "success" response from the server for taking it investigated.  We re-fetch the data every two minutes (iirc) which will catch if another person (or same person in another page) updated something.

File: treeherder/push_health/tests.py [hunk 2]
```diff
@@ -123,6 +123,16 @@ def get_current_test_failures(push, option_map, jobs):
...
+            investigated_test.test == test_name for investigated_test in investigated_tests
+        )
+        is_investigated = False
```
Comment (camd) [inline treeherder/push_health/tests.py:129]: Is this some test code?  Why are we setting `is_investigated = False` here?

Comment (camd) [reply treeherder/push_health/tests.py:129]: Just asking since you're checking for it in line 126 above.  Seems like we're doing double-duty here.

Comment (SuyashSalampuria) [reply treeherder/push_health/tests.py:129]: Ah yes
This is test code
I will just remove it now
Sorry for it

File: treeherder/model/migrations/0021_investigatedtests.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    dependencies = [
+        ('model', '0020_textlogerror_job'),
+    ]
+
+    operations = [
+        migrations.CreateModel(
+            name='InvestigatedTests',
+            fields=[
+                ('id', models.BigAutoField(primary_key=True, serialize=False)),
+                ('test', models.CharField(max_length=255)),
```
Comment (jmaher) [inline treeherder/model/migrations/0021_investigatedtests.py:18]: is this long enough?  I would think we have some edge cases

File: treeherder/push_health/tests.py (hunk 1 above)
Comment (jmaher) [reply treeherder/push_health/tests.py:126]: slight nit; isClassified vs is_investigated, we should keep camelCase

**Comment by camd** [comment]
I mentioned this in Riot, but you follow these instructions to fix that file: https://github.com/mozilla/treeherder/blob/5c3fd90ebb94a892f470d2431333d7260ae3a6ce/tests/extract/test_extract_job.py#L109

File: treeherder/push_health/tests.py [hunk 4]
```diff
@@ -142,6 +152,8 @@ def get_current_test_failures(push, option_map, jobs):
                 'failedInParent': False,
                 'passFailRatio': passFailRatio,
                 'isClassifiedIntermittent': isClassifiedIntermittent,
+                'isInvestigated': is_investigated,
+                'investigatedTestId': investigated_test_id,
```
Comment (jmaher) [inline treeherder/push_health/tests.py:156]: please change to camelCase for both isInvestigated and investigatedTestId

**Comment by camd** [comment]
We should probably get better instructions in our docs for updating those files.  :)

File: treeherder/push_health/tests.py [hunk 5]
```diff
@@ -209,12 +221,13 @@ def get_test_failures(
...
         2, push_date, fixed_by_commit_history_days, option_map, repository_ids
     )
+    investigated_tests = InvestigatedTests.objects.filter(push=push)
 
     # ``push_failures`` are tests that have FailureLine records created by out Log Parser.
     #     These are tests we are able to show to examine to see if we can determine they are
     #     intermittent.  If they are not, we tell the user they need investigation.
     # These are failures ONLY for the current push, not relative to history.
-    push_failures = get_current_test_failures(push, option_map, jobs)
+    push_failures = get_current_test_failures(push, option_map, jobs, investigated_tests)
```
Comment (jmaher) [inline treeherder/push_health/tests.py:230]: I like this here; seems extensible

File: ui/push-health/PlatformConfig.jsx [hunk 6]
```diff
@@ -56,11 +70,19 @@ class PlatformConfig extends React.PureComponent {
 
     return (
       <Row
-        className="ml-5 pt-2 mr-1"
+        className={`ml-5 pt-2 mr-1 ${isInvestigated && 'investigated'}`}
```
Comment (jmaher) [inline ui/push-health/PlatformConfig.jsx:73]: I don't understand the need for:
&& 'investigated' ?  does isInvestigated have a string, not a boolean?

**Comment by camd** [comment]
Oh, and make sure if the user is NOT logged in at the time to either kick off logging them in, or at least put up a `notify` message telling them they need to log in.

File: ui/push-health/PlatformConfig.jsx (hunk 6 above)
Comment (camd) [reply ui/push-health/PlatformConfig.jsx:73]: This is a way only add that text (in this case the class of `investigated` if the `isInvestigated` value is set to true.  It's just a syntactic shortcut.

File: treeherder/model/migrations/0021_investigatedtests.py (hunk 3 above)
Comment (camd) [reply treeherder/model/migrations/0021_investigatedtests.py:18]: Good point, yeah.  Maybe 350 would be better?

**Comment by camd** [comment]
I can confirm that it is waiting to update the UI until it polls for new data from the server.  This is done in the `Health.jsx` `updatePushHealth()` function every two minutes.  

It occurs to me that you could just trigger a call to `updatePushHealth` whenever you mark as investigated or uninvestigated.in the `investigateTest` and `uninvestigateTest` functions.

I'm noticing that it seems to match too much when a test is marked investigated.  I only marked one of them investigated, but it shows all 4 of these crossed off:

Screen Shot 2020-08-20 at 9 08 55 AM

**Comment by camd** [comment]
Also, after changing `investigated` status, it should un-check the item(s) that just changed.

File: treeherder/model/migrations/0021_investigatedtests.py (hunk 3 above)
Comment (SuyashSalampuria) [reply treeherder/model/migrations/0021_investigatedtests.py:18]: I have made it a textfield because the corresponding things in the jobType model were TextField
Should I change it

**Comment by camd** [comment]
> > Also, after changing investigated status, it should un-check the item(s) that just changed.
> 
> I had a thought in implementing this
> I used the Gmail's mark as read feture which is similar to what we are implementing
> The selected mails are not unselected after we mark them read or unread
> So, I left it out actually
> I will do this too. Will be just small change

Ahh, I see.  Well, it "feels" like it would be better to clear our check-marks.  Maybe @sarah-clements has an opinion on that?

**Comment by sarah-clements** [comment]
> > > Also, after changing investigated status, it should un-check the item(s) that just changed.
> > 
> > 
> > I had a thought in implementing this
> > I used the Gmail's mark as read feture which is similar to what we are implementing
> > The selected mails are not unselected after we mark them read or unread
> > So, I left it out actually
> > I will do this too. Will be just small change
> 
> Ahh, I see. That makes sense. We could go either way on it. But somehow it "feels" like it would be better to clear our check-marks. Maybe @sarah-clements has an opinion on that?

Yeah, I think we should have them clear after action is performed. I have a vague recollection one of our test users expected that to happen during a demo.

**Comment by camd** [comment]
I will take a look at the latest changes today.

**Comment by SuyashSalampuria** [comment]
> I will take a look at the latest changes today.

Thanks a lot.
There has been some issue with black
Other tests are passing

**Comment by camd** [comment]
I made some progress looking into this today, but am not done.  I will continue evaluating it tomorrow.

**Comment by camd** [comment]
I apologize.  I didn't get a chance to work on this today.  Other things bubbled to the top of my list.

**Comment by camd** [comment]
I'm reviewing this now.  I'm ingesting data locally so I can test it, so that's taking some time... :)

File: treeherder/model/models.py [hunk 7]
```diff
@@ -1402,3 +1402,16 @@ class Meta:
...
+    """Tests that have been marked as investigated in the Push health UI.  These act like
+    completed To Do List items."""
+
+    id = models.BigAutoField(primary_key=True)
+    job_type = models.ForeignKey(JobType, related_name="investigated", on_delete=models.CASCADE)
+    test = models.CharField(max_length=255)
+    push = models.ForeignKey(Push, on_delete=models.CASCADE)
+
+    class Meta:
+        unique_together = ["job_type", "test", "push"]
```
Comment (camd) [inline treeherder/model/models.py:1417]: Let's also add:

            db_table = 'investigated_tests'

Otherwise, it will create the table as: `model_investigatedtests` which does not match our naming scheme.

File: treeherder/push_health/tests.py [hunk 8]
```diff
@@ -124,6 +124,14 @@ def get_current_test_failures(push, option_map, jobs):
             job['failure_classification_id'] == 4 for job in jobs[job_name]
         )
 
+        isInvestigated = False
+        investigatedTestId = None
+        for investigatedTest in investigatedTests:
+            if investigatedTest.test == test_name:
```
Comment (camd) [inline treeherder/push_health/tests.py:130]: This should also be checking that the job_type_id matches.  Currently, it marks all failures on all platforms for this test as investigated, when only one was checked to be marked.

File: ui/push-health/PlatformConfig.jsx [hunk 9]
```diff
@@ -88,11 +107,21 @@ class PlatformConfig extends React.PureComponent {
         style={{ background: '#f2f2f2' }}
       >
         <Row className="ml-2 w-100 mb-2 justify-content-between">
+          <Col xs="auto">
+            <Input
```
Comment (camd) [inline ui/push-health/PlatformConfig.jsx:111]: Is it possible to make the check box a bit larger?  They're pretty small on my screen:
Screen Shot 2020-09-28 at 3 32 02 PM


Maybe something along the lines of:

    width: 30px;
    height: 30px;
    margin-top: 0 !important;

Though that isn't quite vertically centered all the way.

File: ui/push-health/Test.jsx [hunk 10]
```diff
@@ -26,6 +48,102 @@ class Test extends PureComponent {
...
+    const { notify, currentRepo, revision, updatePushHealth } = this.props;
+
+    let data;
+    let failureStatus;
+    if (selectedTests.size === 0) {
+      notify(`Select atleast one test`, 'warning');
+    } else {
+      selectedTests.forEach(async (test) => {
+        ({ data, failureStatus } = await create(
+          `${getProjectUrl(
```
Comment (camd) [inline ui/push-health/Test.jsx:93]: It would be more efficient to save this url outside the for-loop so you don't create it each time you loop.

File: ui/push-health/Test.jsx [hunk 11]
```diff
@@ -26,6 +48,102 @@ class Test extends PureComponent {
...
+    const { selectedTests } = this.state;
+    const { notify, currentRepo, revision, updatePushHealth } = this.props;
+
+    if (selectedTests.size === 0) {
+      notify(`Select atleast one test`, 'warning');
+    } else {
+      selectedTests.forEach(async (test) => {
+        if (test.isInvestigated) {
+          const response = await destroy(
+            `${getProjectUrl(
```
Comment (camd) [inline ui/push-health/Test.jsx:125]: Same here with generating the URL just once outside the for loop.

File: treeherder/model/migrations/0021_investigatedtests.py (hunk 3 above)
Comment (camd) [reply treeherder/model/migrations/0021_investigatedtests.py:18]: Sure, textfield sounds good.

**Review by camd — CHANGES_REQUESTED** [review]
This is looking pretty good.  A few bugs as mentioned inline.  This could use a test in the python code to be sure that only the right tests are marked as investigated.

Thank you!

File: ui/push-health/PlatformConfig.jsx (hunk 9 above)
Comment (SuyashSalampuria) [reply ui/push-health/PlatformConfig.jsx:111]: Achieved this using transform scale(1.2)

File: ui/push-health/Test.jsx (hunk 11 above)
Comment (SuyashSalampuria) [reply ui/push-health/Test.jsx:123]: We are requesting on a new URL each time
Since the test Id is change.
So, this has to be sone

**Comment by SuyashSalampuria** [comment]
I have made all the changes @camd 
Theres some trouble seems to be with Black on my system.
Rest of the things are running fine now

**Comment by camd** [comment]
Thanks!  Taking a look at this now.

**Review by camd — APPROVED** [review]
I fixed up the `black` issues.  All good now.  Thanks for all the great changes!  I'll merge now.

**Comment by SuyashSalampuria** [comment]
Thanks a  lot @camd :)
